### PR TITLE
Added Features

### PR DESCRIPTION
**Pull request by olliklee**
Added Judoon (Doctor Who Alien Language)
Added Ragbaby cypher. Was asked in issues.

### CODE REVIEW

File: lib/application/i18n/assets/de.json [hunk 1]
```diff
@@ -3172,6 +3172,9 @@
   "irrationalnumbers_search_title": "Suche Vorkommen",
   "irrationalnumbers_error_nonumbers": "Bitte mindestens eine Ziffer eingeben.",
   "irrationalnumbers_search_description": "Durchsucht die ersten 1 000 000 Nachkommastellen auf alle Vorkommen einer Zahlengruppe",
+  "judoon_title": "Judoon Übersetzer",
```
Comment (S-Man42) [inline lib/application/i18n/assets/de.json:3175]: Please only add en.json. The translations will be done in external tool Crowdin. To avoid certain conflicts, please remove the changes here.

File: lib/application/i18n/assets/en.json [hunk 2]
```diff
@@ -3171,6 +3171,9 @@
   "irrationalnumbers_search_title": "Search Occurrence",
   "irrationalnumbers_error_nonumbers": "Please enter at least one number.",
   "irrationalnumbers_search_description": "Searches all occurences of number groups within the first 1 000 000 digits",
+  "judoon_title": "Judoon's Language",
+  "judoon_description": "Each letter is spelled with a special 'o' loud ; inspired by 'Doctor Who'",
```
Comment (S-Man42) [inline lib/application/i18n/assets/en.json:3175]: Just a small thing: Please remove the space before the semicolon

File: lib/application/i18n/assets/en.json [hunk 3]
```diff
@@ -4004,6 +4007,9 @@
   "rabbit_description": "(Secret) Rabbit is a high-speed stream cipher",
   "rabbit_initialization_vector": "Initialization Vector (Optional)",
   "rabbit_initialization_vector_format": "Initialization vector type does not match the initialization vector!",
+  "ragbaby_title": "Ragbaby Cypher",
```
Comment (S-Man42) [inline lib/application/i18n/assets/en.json:4010]: Please leave "Cypher" out from the title for consistent naming :)

File: lib/application/i18n/assets/en.json [hunk 4]
```diff
@@ -4004,6 +4007,9 @@
...
+  "ragbaby_description": "This is substitution cypher with a 24 letters alphabet. J=>I and X=>U. The rotation changes for each word. \n More under https://www.cwu.edu/academics/math/_documents/kryptos-challenges/cwu-kryptos-rag-baby-cipher.pdf",
```
Comment (S-Man42) [inline lib/application/i18n/assets/en.json:4011]: Please substitute the => with a real arrow character. See different examples. Please remove the second part completely. We will habe a separate manual page after the release. Not necessary to link other sources, especially if the url is not clickable. If source must be named for license purposes, please use the license.dart file instead.

File: lib/application/i18n/assets/en.json [hunk 5]
```diff
@@ -4004,6 +4007,9 @@
...
+  "ragbaby_example": "Hello World -> igopt zrvqk",
```
Comment (S-Man42) [inline lib/application/i18n/assets/en.json:4012]: Arrow character again :)

File: lib/application/registry.dart [hunk 6]
```diff
@@ -941,6 +943,11 @@ void initializeRegistry(BuildContext context) {
       'rotation',
       'kamasutra',
     ]),
+    GCWTool(tool: const Judoon(), id: 'judoon', categories: const [
+      ToolCategory.CRYPTOGRAPHY
+    ], searchKeys: const [
```
Comment (S-Man42) [inline lib/application/registry.dart:948]: Please remove the "doctor who" search key here. The registry only holds references to the search files (in the assets as well). So if doctor who is necessary, put this there as search word under the key 'judoon' which you define here.

File: lib/application/registry.dart [hunk 7]
```diff
@@ -1145,6 +1152,11 @@ void initializeRegistry(BuildContext context) {
     ], searchKeys: const [
       'rabbit',
     ]),
+    GCWTool(tool: const Ragbaby(), id: 'ragbaby', categories: const [
+      ToolCategory.CRYPTOGRAPHY
+    ], searchKeys: const [
+      'ragbaby',
```
Comment (S-Man42) [inline lib/application/registry.dart:1158]: Same. here. You define a search key, but did not add one in the search files :)

File: lib/tools/crypto_and_encodings/judoon/logic/judoon.dart [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      .split('')
+      .map((char) => alphabetJudoon[char.toLowerCase()] ?? '')
+      .join(' ');
+}
+
+String decryptJudoon(String input) {
+  if (input.isEmpty) return "";
+
+  final invertedDict = {
+    for (var entry in alphabetJudoon.entries) entry.value: entry.key
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/judoon/logic/judoon.dart:35]: There is already a common function for this call. switchKeyValueMap (or similar). Please try to reuse common functions, of possible. If it seems to be a common operation, the chance is good, that here ia already something written and tested. If not, feel free to add such general code into common utils

File: lib/tools/crypto_and_encodings/judoon/widget/judoon.dart [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      children: <Widget>[
+        GCWTextField(
+          controller: _inputController,
+          onChanged: (text) {
+            setState(() {
+              _currentInput = text;
+            });
+          },
+        ),
+        GCWTextDivider(text: i18n(context, 'common_key')),
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/judoon/widget/judoon.dart:48]: Without having seen the tool currently, I guess, it is is not necessary to add a divider just for one switch widget. The switch has a title property for this.

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart [hunk 10]
```diff
@@ -0,0 +1,85 @@
+import 'package:gc_wizard/utils/alphabets.dart';
+import 'package:gc_wizard/utils/string_utils.dart';
+
+
+String translatePasswordIntoKey(String input) {
+  List<String> alphabet = alphabet_AZString.split("");
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:6]: Nothing wrong, but common sense in the project
: For string we are always using the '  instead " ... Please change to stay consistent throughout the project.

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  alphabet.remove("X");
+
+  String newAlphabet = "";
+
+  if (input.isEmpty) {
+    return alphabet.join();
+  } else {
+    String cleanedInput = trimCharacters(input, " ")
+        .toUpperCase();
+
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:17]: What happens it the input is empty after trimming? Doesn't it makes sense to return then? So your if could be after trimming.

Comment (S-Man42) [reply lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:17]: And as far as I can see, it is not necessary to put everything into an else branch.To keep the code clean, you can remove this. It avoids unnecessary nestings and stays more readable. So ask for empty alphabet (after trim) and if so, return, if not go further. no need for else here.

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart [hunk 12]
```diff
@@ -0,0 +1,85 @@
+import 'package:gc_wizard/utils/alphabets.dart';
+import 'package:gc_wizard/utils/string_utils.dart';
+
+
+String translatePasswordIntoKey(String input) {
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:5]: Please try to make everything private. In Dart you need to put the underscore before the variables, functions, whatever, names. So here you need to name it _translatePasswordIntoKey. Please check entire code for this, I haven't had a look onzo this so far.

File: lib/tools/crypto_and_encodings/ragbaby/widget/ragbaby.dart [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+          hintText: i18n(context, 'common_key'),
+          controller: _passwordController,
+          onChanged: (text) {
+            setState(() {
+              _currentPassword = text;
+            });
+          },
+        ),
+        const GCWDivider(),
+        GCWText(text: i18n(context, "common_alphabet") + ": " + translatePasswordIntoKey(_currentPassword)),
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/widget/ragbaby.dart:63]: See hint from above: As long as you don't create sections with several widgets or you don't need very long texts, please. try to use the title property instead og a divider. It saves space.

**Comment by S-Man42** [comment]
All in all, it looks fine at the first glance. Greatly done!

However, I did not run it so far, which I will do the next days. And I will have more detailed look. at the code when I am not only on my smartphone :D

Well

1. Please add the search files. For writing correct search strings, please read the manual:  https://blog.gcwizard.net/de/2022/01/16/schreiben-von-suchtexten/. Here it is as said above for normal i18n pages: Please fill out only the en.jaon file, the translations can be done furtherly on Crowdin, when we uploaded the data
2. It is absolutely necessary, that you write proper tests for every kind of logic you check in (see the test directory). So. please add logic for your both functions.

File: lib/application/i18n/assets/en.json [hunk 14]
```diff
@@ -3171,6 +3171,9 @@
   "irrationalnumbers_search_title": "Search Occurrence",
   "irrationalnumbers_error_nonumbers": "Please enter at least one number.",
   "irrationalnumbers_search_description": "Searches all occurences of number groups within the first 1 000 000 digits",
+  "judoon_title": "Judoon's Language",
```
Comment (S-Man42) [inline lib/application/i18n/assets/en.json:3174]: I would simply name it Judoon. I believe it fits more into the entire app look and feel...

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart (hunk 10 above)
Comment (olliklee) [reply lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:6]: Everywhere?

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart (hunk 12 above)
Comment (olliklee) [reply lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:5]: I use that function in /widget/ragbaby.dart, too.
If I make it private, I can't use the alphabet string there. 
Or is there another way?

**Comment by olliklee** [comment]
I'm gonna add testing later.
Thanks for all the comments. Lots of stuff to learn 🤯 😄

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart (hunk 10 above)
Comment (S-Man42) [reply lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart]: Yep.

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    // tranlates first or only character
+    if (newPattern.isNotEmpty) {
+      var firstChar = newPattern[0];
+      if (translation.containsKey(firstChar)) {
+        output += translation[firstChar]!;
+      }
+      newPattern = newPattern.substring(1);
+    }
+    // recursion
+    return _translate(newPattern, output);
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/logic/major.dart:75]: A recursion looks extremely uneffecient. Why not one single loop over the "special translations" and afterwards one over the letters left? The entire code might look like this:

```
var output = substition(input, specialTranslations);
return subsitution(output, translations);
```

(even without the internal substitution function this would be only the two loops. A recursion is always nasty to debug and a lot more inefficient due to the increasing working memory consumption)

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  // Remove all unnecessary chars like puctuation and diacritics and
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/logic/major.dart:78]: typo in punctuation

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+  // returns a single string with spaces between single words
+  String _cleanText() {
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/logic/major.dart:80]: This is can be done much easier as I see it:

var output = removeDiacritics(text.toUppertext()); // only non-diacritics, only capital letters
return output.replaceAll(RegExp(r'[^A-Z]'), ''); // remove all characters, that are not A-Z

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+      }
+      newText = validWords;
+    }
+    return newText.join(' ').toLowerCase();
+  }
+
+  // returns a list of consonant groups
+  // like [d, r, m, nd, g, ht, m, s, chs]
+  List<String> _groups() {
+    var words = _cleanText().split(" ");
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/logic/major.dart:107]: please remember the " vs ' issue :)

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+        }
+      }
+      newText = validWords;
+    }
+    return newText.join(' ').toLowerCase();
+  }
+
+  // returns a list of consonant groups
+  // like [d, r, m, nd, g, ht, m, s, chs]
+  List<String> _groups() {
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/logic/major.dart:106]: This is not clear to me, what should this do?

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+    List<String> groupList = [];
+    for (var word in words) {
+      var groups = word.split(RegExp('[aeiouqxy]+'));
+      groupList.addAll(groups);
+    }
+    groupList.removeWhere((entry) => entry.isEmpty); // remove empty entries
+    return groupList;
+  }
+
+  String _replaceDoubleLetters(String input) {
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/logic/major.dart:117]: I am not quite sure, if there isn't already something similar in the common utils. If not, this is a good candidate for the string utils class.

File: lib/tools/crypto_and_encodings/major/widget/major.dart [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+        GCWTextField(
+          controller: _inputController,
+          onChanged: (text) {
+            setState(() {
+              _currentInput = text;
+            });
+          },
+        ),
+        
+        GCWTextDivider(text: i18n(context, 'major_settings_capitalized_only')),
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/widget/major.dart:48]: Please don't you a separate divider widget if the the "title" parameter of the widget is enough. Please always save space where possible.

File: lib/tools/crypto_and_encodings/major/widget/major.dart [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+          leftValue: i18n(context, 'common_no'),
+          rightValue: i18n(context, 'common_yes'),
+          value: _nounMode,
+          onChanged: (value) {
+            setState(() {
+              _nounMode = value;
+            });
+          },
+        ),
+        GCWTextDivider(text: i18n(context, 'major_output_plaintext')),
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/widget/major.dart:60]: Every output should go into the Output section which begins with the GCWDefaultOutput widget in your case. What is that meant to be here? (I still cannot compile, so I can't have a look)

File: lib/tools/crypto_and_encodings/major/widget/major.dart [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+        GCWText(text: MajorDecrypt(_currentInput).toString()),
+
+        GCWDefaultOutput(child: _buildOutput())
+      ],
+    );
+  }
+
+  String _buildOutput() {
+    return (_nounMode == GCWSwitchPosition.left)
+        ? MajorDecrypt(_currentInput).decodeToNumberString()
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/major/widget/major.dart:70]: I would shorten it to return MajorDecrypt(_currentInput, nounMode: noneMode == GCWSwitchPosition.left).decodeToNumberString()

File: lib/tools/crypto_and_encodings/ragbaby/widget/ragbaby.dart [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        GCWTextField(
+          hintText: i18n(context, 'common_key'),
+          controller: _passwordController,
+          onChanged: (text) {
+            setState(() {
+              _currentPassword = text;
+            });
+          },
+        ),
+        GCWTextDivider(text: i18n(context, "common_alphabet")),
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/widget/ragbaby.dart:62]: Still " chars.

Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/widget/ragbaby.dart:62]: Please remember: Use TextDividers as few as possible. It is not only about space, but also about visual aspects. Too much horizontal lines simply look weird...

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+  List<String> alphabet = alphabet_AZString.split('');
+  alphabet.remove('J');
+  alphabet.remove('X');
+
+  if (input.isEmpty) return alphabet.join();
+
+  String cleanedInput = input.toUpperCase();
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:15]: Are you sure that simply uppercasing is enough for sanitizing here? I can't see the entire context atm, so I am not quite sure. Just asking

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart [hunk 26]
```diff
@@ -0,0 +1,80 @@
+import 'package:gc_wizard/utils/alphabets.dart';
+
+//
+// More under
+// https://www.cwu.edu/academics/math/_documents/kryptos-challenges/cwu-kryptos-rag-baby-cipher.pdf
+
+
+String translatePasswordIntoKey(String input) {
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:8]: Please make it private

File: lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+
+String encryptRagbaby(String input, String password) {
+  if (input.isEmpty) return '';
+
+  String key = translatePasswordIntoKey(password);
+  List<String> cleanedText =
+  input.toUpperCase().replaceAll('X', 'U').replaceAll('J', 'I').split(' ');
+  List<String> encryptedText = [];
+
+  for (int wIndex = 0; wIndex < cleanedText.length; wIndex++) {
```
Comment (S-Man42) [inline lib/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart:36]: Not a big thing, but please have a look at the map() function for Dart. In my opinions this could have done much cleaner with split(' ').map(...).join('');

File: test/tools/crypto_and_encodings/judoon/logic/judoon-test.dart [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+import 'package:gc_wizard/tools/crypto_and_encodings/judoon/logic/judoon.dart';
+
+void main() {
+  group("Judoon.encryptJudoon:", () {
+    List<Map<String, Object?>> _inputsToExpected = [
+      {'input' : '', 'expectedOutput' : ''},
+
+      {'input' : 'quick! brown fox ("jumps").', 'expectedOutput' : 'qwo tro kro co kno sco sho bo ro plo wo no sho fo plo xo sho pra cho jo tro mo po so cho pla bla'},
+      {'input' : '1234567890', 'expectedOutput' : 'ha ta tra fa fla sa schla ga na za'},
+      {'input' : 'QUICK BROWN FOX äöüß', 'key': 8, 'expectedOutput' : 'qwo tro kro co kno sho bo ro plo wo no sho fo plo xo sho • • • •'},
```
Comment (S-Man42) [inline test/tools/crypto_and_encodings/judoon/logic/judoon-test.dart:11]: Are the dots at the end part of the code or did you put them just for invalid letters? If second: Please don't do this. We always sanitize the input that way, that invalid input will be made valid before encode/decoding. So there will be always a valid output.

File: test/tools/crypto_and_encodings/ragbaby/logic/ragbaby-test.dart [hunk 29]
```diff
@@ -0,0 +1,40 @@
+import "package:flutter_test/flutter_test.dart";
+import 'package:gc_wizard/tools/crypto_and_encodings/ragbaby/logic/ragbaby.dart';
+
+void main() {
+  group("Ragbaby.encryptRagbaby:", () {
+    List<Map<String, Object?>> _inputsToExpected = [
+      {'input' : '', 'pw': '', 'expectedOutput' : ''},
+
+      {'input' : 'The quick brown fox jumps.', 'pw': 'its a secret', 'expectedOutput' : 'SLB VYEGQ GGWAY LWT CTWTH.'},
```
Comment (S-Man42) [inline test/tools/crypto_and_encodings/ragbaby/logic/ragbaby-test.dart:9]: more test cases are possible: Non-empty input, but empty pw. Or empty input but non-empty pw. PW with special characters. Input with numbers, ...

File: lib/application/i18n/assets/en.json (hunk 3 above)
Comment (S-Man42) [reply lib/application/i18n/assets/en.json:4016]: "Cypher" is still part of the string...

File: lib/application/i18n/assets/en.json (hunk 4 above)
Comment (S-Man42) [reply lib/application/i18n/assets/en.json:4017]: Well, it seems you marked the conversation as resolved whereas the "problems" are still there... Right? Did you missed to merge fixes?

File: lib/application/searchstrings/assets/en.json [hunk 30]
```diff
@@ -259,6 +259,7 @@
   "iteratedcrosssumrange": "range",
   "iteratedcrossumrangefrequency": "frequency  occurrences",
   "japanese_numerals": "numbers digits numerals japanese",
+  "judoon": "doctor who",
```
Comment (S-Man42) [inline lib/application/searchstrings/assets/en.json:262]: write together. So you could find both "Doctorwho" and "Docter Who" input. ;)

File: lib/application/searchstrings/assets/en.json [hunk 31]
```diff
@@ -271,6 +272,7 @@
   "lcm": "lcm leastcommonmultiplier",
   "licenses": "licenses licences library libraries",
   "magic_eye": "3dstereograms autostereograms 3dpictures 3dimages",
+  "major": "majorsystem phoneticnnumbersystem phoneticmnemonicsystem herigonesmnemonicsystem",
```
Comment (S-Man42) [inline lib/application/searchstrings/assets/en.json:275]: According to Wikipedia I would add: major-system, aimeeparis, aimeparis, mastersystem

File: lib/application/searchstrings/assets/en.json (hunk 30 above)
Comment (S-Man42) [reply lib/application/searchstrings/assets/en.json:262]: Add judoon as well

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 32]
```diff
@@ -0,0 +1,125 @@
...
+    'r': '4',
+    'l': '5',
+    'j': '6',
+    'g': '7',
+    'k': '7',
+    'c': '7',
+    'f': '8',
+    'v': '8',
+    'w': '8',
+    'ph': 'f',
```
[the remark this answers is not in the document]
Comment (olliklee) [reply lib/tools/crypto_and_encodings/major/logic/major.dart:27]: Changed

File: lib/tools/crypto_and_encodings/major/logic/major.dart [hunk 33]
```diff
@@ -0,0 +1,125 @@
+import 'package:diacritic/diacritic.dart';
+
+class MajorDecrypt {
+  String text;
+  bool nounMode;
+
+  MajorDecrypt(this.text, {this.nounMode = false});
+
+  final _noLetterChars = RegExp(r'[ !"#$%&()*+,-./:;<=>?@[\\\]^_`{|}~\s]+');
```
[the remark this answers is not in the document]
Comment (olliklee) [reply lib/tools/crypto_and_encodings/major/logic/major.dart:9]: This is meant to split words and sentences.

File: test/tools/crypto_and_encodings/ragbaby/logic/ragbaby-test.dart (hunk 29 above)
Comment (olliklee) [reply test/tools/crypto_and_encodings/ragbaby/logic/ragbaby-test.dart:9]: Unknown characters in input are just copied at the same place in the output.
Unknown characters in key are just ignored.
If password is empty, the key alphabet is automatically replaced with the untouched alphabet.
That's why I display the alphabet, which is used for the encryption.

File: lib/application/searchstrings/assets/en.json (hunk 31 above)
Comment (olliklee) [reply lib/application/searchstrings/assets/en.json:275]: the words I used were from the English wiki entry. I add those words as, too
